src/core/execute.c

```diff
@@ -3250,11 +3267,22 @@ static int exec_child(
                 }
         }
 
-        if (context->cpu_set.set)
-                if (sched_setaffinity(0, context->cpu_set.allocated, context->cpu_set.set) < 0) {
+        if (context->cpu_set.set || context->derive_cpu_affinity_from_numa_mask) {
+                _cleanup_(cpu_set_reset) CPUSet mask = {};
+
+                cpu_set_add_all(&mask, &context->cpu_set);
+
+                r = exec_context_cpu_affinity_from_numa(context, &mask);
+                if (r < 0) {
+                        *exit_status = EXIT_CPUAFFINITY;
+                        return log_unit_error_errno(unit, r, "Failed to derive CPU affinity mask from NUMA mask: %m");
+                }
```

Comment:
hmm, if `derive_cpu_affinity_from_numa_mask` is true I'd really not bother with `context->cpud_set`...


maybe:

```c
_cleanup_(cpu_set_reset) CPUSet converted_cpu_set = {};
const CPUSet *cpu_set = NULL;

…

if (context->dervice_cpu_affinity_from_numa_mask) {
        r = numa_to_cpu_set(context->numa_policy, &converted_cpu_set);
        if (r < 0) {
                *exit_status = EXIT_CPUAFFINITY;
                return log_unit_error_errno(…, "Failed to convert NUMA policy to CPU mask: %m");
        }
        cpu_set = &converted_cpu_set;
} else if (context->cpu_set.set) 
        cpu_set = &context->cpu_set;

if (cpu_set)
        if (sched_setaffinity(0,cpu_set->allocated, cpu_set->set) < 0) {
                …
        }
```

I think the intention becomes much clearer then: if conversion is requested then convert, if not, use the explicit one if it is set.